src/core/dependency-graph.ts

```diff
@@ -39,20 +87,51 @@ export class DependencyGraph {
    * @param path The path (i.e. url) of a document.
    * @param newDependencies The paths of that document's direct dependencies.
    */
-  addDependenciesOf(path: string, newDependencies: Iterable<string>) {
-    let dependencies = this._dependencies.get(path);
-    if (!dependencies) {
-      dependencies = new Set();
-      this._dependencies.set(path, dependencies);
+  addDocument(url: string, dependencies: Iterable<string>) {
+    const record = this._getRecordFor(url);
+    for (const dependency of dependencies) {
+      record.dependencies.add(dependency);
+      const dependantRecord = this._getRecordFor(dependency);
+      dependantRecord.dependants.add(url);
+    }
+    try {
+      record._dependenciesDeferred.resolve(record.dependencies);
+    } catch (e) {
+      console.log(`error adding ${url}`);
+      throw e;
+    }
+  }
+
+  rejectDocument(url: string, error: Error) {
+    this._getRecordFor(url)._dependenciesDeferred.reject(error);
+  }
+
+  /**
+   * Returns a Promise that resolves when the given document and all
+   * of its transitive dependencies have been resolved or rejected. This
+   * Promise never rejects, if the document or any dependencies are rejected,
+   * the Promise still resolves.
+   */
+  async whenReady(url: string): Promise<void> {
+    await this._whenReady(url, new Set<string>());
+  }
+
+  private async _whenReady(key: string, visited: Set<string>) {
+    if (visited.has(key)) {
+      return;
     }
-    for (const newDependency of newDependencies) {
-      dependencies.add(newDependency);
-      let dependants = this._dependants.get(newDependency);
-      if (!dependants) {
-        dependants = new Set();
-        this._dependants.set(newDependency, dependants);
+    visited.add(key);
+    try {
+      const deps = await this._getRecordFor(key).dependenciesKnown;
+      for (const dep of deps) {
+        try {
+          await this._whenReady(dep, visited);
+        } catch (ee) {
+          // ignore rejected dependencies
+        }
       }
-      dependants.add(path);
+    } catch (e) {
+      // turn rejection into resolution
```

Comment:
The outer try was needed to catch per-file rejections, but the inner wasn't since this function only called itself. I redid this to use a catch and added some intermediate vars to make it more clear.